source/adapters/level_zero/command_buffer.cpp

```diff
@@ -170,6 +170,64 @@ ur_result_t getEventsFromSyncPoints(
   return UR_RESULT_SUCCESS;
 }
 
+/**
+ * If necessary, it creates a signal event and appends it to the previous
+ * command list (copy or compute), to indicate when it's finished executing.
+ * @param[in] CommandType The type of the command.
+ * @param[in] CommandBuffer The CommandBuffer where the command is appended.
+ * @param[in] isFirstNode A boolean inidicating if the current node is at the
+ * beginning of the command list.
+ * @param[in] isCopy A boolean indicating if the current command uses copy
+ * engine.
+ * @param[out] ZeSignalPrevCommandEvent The event which signals when the
+ * previous command appended to a different command list is finished executing.
+ * @return UR_RESULT_SUCCESS or an error code on failure
+ */
+ur_result_t createSyncPointBetweenCopyAndCompute(
+    ur_command_t CommandType, ur_exp_command_buffer_handle_t CommandBuffer,
+    bool isFirstNode, bool isCopy,
+    ze_event_handle_t &ZeSignalPrevCommandEvent) {
+  if (isFirstNode) {
+    CommandBuffer->MCopyCommandListEmpty = !isCopy;
+    CommandBuffer->MComputeCommandListEmpty = isCopy;
+    CommandBuffer->WasPrevCopyCommandList = isCopy;
+    return UR_RESULT_SUCCESS;
+  } else if (CommandBuffer->WasPrevCopyCommandList == isCopy) {
+    return UR_RESULT_SUCCESS;
+  }
+
+  ur_event_handle_t SignalPrevCommandEvent = nullptr;
+  UR_CALL(EventCreate(CommandBuffer->Context, nullptr /*Queue*/,
+                      false /*IsMultiDevice*/, false, &SignalPrevCommandEvent,
+                      false /*CounterBasedEventEnabled*/,
+                      !CommandBuffer->IsProfilingEnabled,
+                      false /*InterruptBasedEventEnabled*/));
+  SignalPrevCommandEvent->CommandType = CommandType;
+
+  /*
+   * If the current CommandType is different from the one used lastly, we
+   * need to append ZeSignalPrevCommandEvent to the latter to know when it's
+   * finished executing.
+   */
+  ZeSignalPrevCommandEvent = SignalPrevCommandEvent->ZeEvent;
+  if (CommandBuffer->WasPrevCopyCommandList && !isCopy) {
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
+               (CommandBuffer->ZeCopyCommandList, ZeSignalPrevCommandEvent));
+    CommandBuffer->WasPrevCopyCommandList = false;
+  } else {
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
+               (CommandBuffer->ZeComputeCommandList, ZeSignalPrevCommandEvent));
+    CommandBuffer->WasPrevCopyCommandList = true;
+  }
+
+  // Get sync point and register the event with it.
+  ur_exp_command_buffer_sync_point_t SyncPoint =
+      CommandBuffer->getNextSyncPoint();
+  CommandBuffer->registerSyncPoint(SyncPoint, SignalPrevCommandEvent);
```

Comment:
Is every call to this function inside a check for `CommandBuffer->IsInOrderCmdList `? If so I don't think we need to care about sync points (and should also stick that in function comment) Also preference for renaming the function to remove "syncpoint" in that case.

Reply:
I'm not sure I follow. `createSyncPointBetweenCopyAndCompute` is only ever called inside `CommandBuffer->IsInOrderCmdList` check to synchronize between the potential `in-orderCopy` vs `in-orderCompute` race. 

AFAIK Registering `SignalPrevCommandEvent` adds it to the event list which is later being reset. If we managed to use counter-based events, lines `223-226` wouldn't be necessary.

I might be misunderstanding the purpose of `getNextSyncPoint()` and `registerSyncPoint()`, maybe there is a more straightforward way to cache this signal even and later reset it but I thought that's the current way.

Reply:
> I might be misunderstanding the purpose of getNextSyncPoint() and registerSyncPoint(), maybe there is a more straightforward way to cache this signal event and later reset it but I thought that's the current way.

This is more what I was getting at. As was trying to ask if we need this sync point if the UR user doesn't care about sync points. However, you've answered my question by saying that we need to do this to cache the L0 event and reset it later. So I guess my feedback is to update the code comment to append ", so that the event is reset between executions"
```